### DISCUSSION TITLE
`Show` message not being posted in `Widget` when popping back to screen / [Feature] A new message when popping back to screen (specific widget become visible again)

### DISCUSSION THREAD

**Discussion by mzebrak** [Textualize/textual#3521] [Q&A]
Initially, when I read the description of the `Show` message
> Sent when a widget has become visible

I thought that it would allow me to hook on the action of re-showing the widget on the screen below when we exit from another screen (pop_screen), because it appears on the screen again.

There is a logical reason for this behaviour / additional hook to exist and it would simplify many things in multi-screen applications.
That's because there is already `on(Mount)` which can be hooked on when we initially push a new screen with new widgets. 
What I'm talking about is like the thing, only from the other side. 

I also believe that textual should provide as many hooks as possible - this is a very nice feature that allows you to get rid of binding in the code.

Let's consider the case when we have, for example, a list of items on one screen, with a "DELETE" button next to each one. after pressing it, the next screen appears on the stack (let's skip thinking about what should be there and why.. e.g.  the delete action confirmation or some extra data should be provided) - then we want that item to be removed (we could hook on this message (posted when other screen pops, and rebuild the desired  item Widget)

When the action is successful - we would like to inform the screen below us. A great solution seems to be to use `@on` hooks instead of using callbacks, either through dismiss or passing references and calling methods on them. It is less binding.

When we are talking about returning to this screen from several places - instead of placing logic informing this Widget in a strange way if it is placed on the screen below and nested (then I would probably have to `self.app.screen_stack[-2].query(SomeItem)`, because` self.app.query` does not seem to take into account inactive screens or  `dismiss` with callback which is also not the perfect solution) in many places - it is enough that in the widget concerned I will handle `@on(Show)` (maybe even instead of `@on(Mount)`)

**Comment by davep** [comment]
I notice you don't mention why [`ScreenResume`](https://textual.textualize.io/events/screen_resume/) doesn't work in your use case. Is there a reason for that?

**Comment by mzebrak** [comment]
> I notice you don't mention why [`ScreenResume`](https://textual.textualize.io/events/screen_resume/) doesn't work in your use case. Is there a reason for that?

`ScreenResume` is posted only on a `Screen` and not on a `Widget`, so there is need to have `on(ScreenResume)`  logic which will be placed in the screen class and when we actually want the action to be about the widget found on it - we have to additionally inform it. 

so

```python
@on(ScreenResume)
def notify_child_widget_about_resuming() -> None
    self.query_one(SomeWidget).notify_resumed()
```

is not that promising.

**Reply by davep** [reply]
True `ScreenResume` is something that goes to a screen, but that could be the basis of giving you what you want. For example, I imagine you could make a base screen for all your other screens which does something like this:

```python
class MyScreen(Screen):

    @on(ScreenResume)
    def notify_children(self, event: ScreenResume) -> None:
        for child in self.children:
            child.post_message(event)
```

Untested (normally I'd actually write a fully-working example but I'm kind of in the middle of something else), but I think that's the route I'd start looking at if I were inclined to handle things the way you seem to want to.
